pandas/_libs/properties.pyi

```diff
@@ -0,0 +1,13 @@
+from typing import (
+    Callable,
+    Protocol,
+    TypeVar,
+)
+
+_F = TypeVar("_F", contravariant=True)
+_G = TypeVar("_G")
+
+class cache_readonly(Protocol[_F, _G]):
+    def __init__(self, func: Callable[[_F], _G]) -> None: ...
+    def __get__(self, obj: _F, typ) -> _G: ...
```

Comment:
> The return value of __get__ can also be self (is that intended?):

Yes, this happens when you lookup cls.foo on the _class_ rather than on an instance

Reply:
Thanks, I added an overload for this.

The current type stubs do not handle this case https://github.com/pandas-dev/pandas/blob/b9e90dc8acf003588678196e7dc8294114918bb2/pandas/_libs/properties.pyx#L30
If the only solution to cover this case is to return `_G | cache_readonly[_F, _G]`, I am inclined to not cover this case as it would make the annotations less useful (even though it would be correct): `index.right` would then be a union of two types.